File: programming-essentials/pe-04/practice.js
const completedExercisesText = "8";
const totalExercisesText = "10";
const bonusPointsText = "5";

const targetPercentage = 80;

const completedExercises = Number(completedExercisesText);
const totalExercises = Number(totalExercisesText);
const bonusPoints = Number(bonusPointsText);

const inputsAreValid =
    Number.isFinite(completedExercises) &&
    Number.isFinite(totalExercises) &&
    Number.isFinite(bonusPoints) &&
    totalExercises > 0 &&
    completedExercises >= 0 &&
    completedExercises <= totalExercises &&
    bonusPoints >= 0;

if (!inputsAreValid) {
    console.log("Invalid input.");
} else {
    const basePercentage =
        (completedExercises / totalExercises) * 100;

    const finalPercentage = Math.min(
        basePercentage + bonusPoints,
        100
    );

    const reachedTarget = finalPercentage >= targetPercentage;

    console.log(`Completed: ${completedExercises}/${totalExercises}`);
    console.log(`Bonus points: ${bonusPoints}`);
    console.log(`Result: ${finalPercentage}%`);
    console.log(
        reachedTarget
            ? "Target reached."
            : "Target not reached."
    );
}

File: programming-essentials/pe-04/practice.py
completed_exercises_text = "8"
total_exercises_text = "10"
bonus_points_text = "5"

target_percentage = 80

try:
    completed_exercises = float(completed_exercises_text)
    total_exercises = float(total_exercises_text)
    bonus_points = float(bonus_points_text)

    inputs_are_valid = (
        total_exercises > 0
        and completed_exercises >= 0
        and completed_exercises <= total_exercises
        and bonus_points >= 0
    )

    if not inputs_are_valid:
        print("Invalid input.")
    else:
        base_percentage = (
            completed_exercises / total_exercises
        ) * 100

        final_percentage = min(
            base_percentage + bonus_points,
            100
        )

        reached_target = (
            final_percentage >= target_percentage
        )

        print(
            f"Completed: "
            f"{completed_exercises:g}/{total_exercises:g}"
        )

        print(f"Bonus points: {bonus_points:g}")
        print(f"Result: {final_percentage:g}%")

        if reached_target:
            print("Target reached.")
        else:
            print("Target not reached.")

except ValueError:
    print("Invalid input: values must be numbers.")